kernel/examples/table-reader/src/main/java/io/delta/kernel/examples/SingleThreadedTableReader.java

```diff
@@ -109,20 +109,18 @@ private int readData(StructType readSchema, Scan scan, int maxRowCount) throws I
                         Row scanFileRow = scanFileRows.next();
                         FileStatus fileStatus =
                             InternalScanFileUtils.getAddFileStatus(scanFileRow);
+                        CloseableIterator<ColumnarBatch> physicalDataIter =
+                            tableClient.getParquetHandler().readParquetFiles(
+                                singletonCloseableIterator(fileStatus),
+                                physicalReadSchema,
+                                Optional.empty());
                         try (
```

Comment:
OutOfMemory doesn't extend `Exception`. If we have to catch fatal errors like OOM, then it needs to catch the root of all exceptions `Throwable`. Given this is a connector code, it has a choice on how to handle fatal errors. For example, there is a JVM setting to set the exit process on OOM. AFAIK for fatal errors the process just exits and there is no easy recovery.

If the connector wants, it can close the iterator too (Our iterators extend `Closeable` which states that `close` can be called multiple times with no side-effects), but doing that could problems in normal scenarios if the iterator is in use by another iterator that connector got from Kernel. 

For the `initIfRequired` change: was trying to catch any exceptions that occur between `Kernel.transformData` beginning and before returning the new iterator. One way is to put a `try...catch` block around the whole function (didn't prefer this to due an extra indentation). The other way (currently implemented) is to lazily init within the new iterator.

I am wondering if our API definitions should be such that Kernel guarantees closing the iterator given as input as long as there are no fatal exceptions?

For the change here, I can revert back to closing both iterators within the `try..with..resource` block, but I don't see if it matters.